File: starter-packs/python/core/game_state.py
import json
from dataclasses import dataclass, field
from typing import List, Optional
from enum import IntEnum

from core.map_state import Point


@dataclass
class Projectile:
    uid: str    = ''
    pos: Point  = field(default_factory=Point)
    dest: Point = field(default_factory=Point)

    def __str__(self) -> str:
        return json.dumps(self.__dict__)


@dataclass
class Blade:
    start: Point    = field(default_factory=Point)
    end: Point      = field(default_factory=Point)
    rotation: float = 0.0

    def __str__(self) -> str:
        return json.dumps(self.__dict__)


@dataclass
class Coin:
    uid: str    = ''
    value: int  = 0
    pos: Point  = field(default_factory=Point)

    def __str__(self) -> str:
        return json.dumps(self.__dict__)


class PlayerWeapon(IntEnum):
    PlayerWeaponNone    = 0
    PlayerWeaponCanon   = 1
    PlayerWeaponBlade   = 2

    def __str__(self) -> str:
        return self.name

@dataclass
class PlayerInfo:
    name: str                       = ''
    color: int                      = 0
    health: int                     = 0
    score: int                      = 0
    pos: Point                      = field(default_factory=Point)
    dest: Point                     = field(default_factory=Point)
    playerWeapon: PlayerWeapon      = PlayerWeapon.PlayerWeaponNone
    projectiles: List[Projectile]   = field(default_factory=list)
    blade: Blade                    = field(default_factory=Blade)

    def isAlive(self) -> bool:
        return self.health > 0

    def __str__(self) -> str:
        return json.dumps(self.__dict__, default=lambda o: o.__dict__, indent=4)


@dataclass
class GameState:
    current_tick: int           = 0
    current_round: int          = 0
    players: List[PlayerInfo]   = field(default_factory=list)
    coins: List[Coin]           = field(default_factory=list)

    def __str__(self) -> str:
        return json.dumps(self.__dict__, default=lambda o: o.__dict__, indent=4)

File: starter-packs/javascript/core/game_state.js
import { Point } from './map_state.js';

class Projectile {
  constructor(uuid = '', pos = new Point(), dest = new Point()) {
    this.uuid = uuid;
    this.pos = pos;
    this.dest = dest;
  }

  toString() {
    return `{"uid":"${this.uid}","pos":${this.pos.toString()},"dest":${this.dest.toString()}}`;
  }
};

class Blade {
  constructor(start = new Point(), end = new Point(), rotation = 0.0) {
    this.start = start;
    this.end = end;
    this.rotation = rotation;
  }

  toString() {
    return `{"start":${this.start.toString()},"end":${this.end.toString()},"rotation":${this.rotation}}`;
  }
};

class Coin {
  constructor(uid = '', value = 0, pos = new Point()) {
    this.uid = uid;
    this.value = value;
    this.pos = pos;
  }

  toString() {
    return `{"uid":"${this.uid}","value":${this.value},"pos":${this.pos.toString()}}`;
  }
};

const PlayerWeapon = Object.freeze({
  PlayerWeaponNone: 0,
  PlayerWeaponCanon: 1,
  PlayerWeaponBlade: 2,
  toString(value) {
    return Object.keys(this).find(key => this[key] === value) || '';
  }
});

class PlayerInfo {
  constructor(
    name = '',
    color = 0,
    health = 0,
    score = 0,
    pos = new Point(),
    dest = null,
    player_weapon = PlayerWeapon.PlayerWeaponNone,
    projectiles = [],
    blade = null,
  ) {
    this.name = name;
    this.color = color;
    this.health = health;
    this.score = score;
    this.pos = pos;
    this.dest = dest;
    this.player_weapon = player_weapon;
    this.projectiles = projectiles;
    this.blade = blade;
  }

  isAlive() { return this.health > 0; }

  toString() {
    const projectiles_str = this.projectiles.map(p => p.toString()).join(',');
    const blade_str = this.blade ? this.blade.toString() : null;

    return `{
      "name":"${this.name}",
      "color":${this.color},
      "health":${this.health},
      "score":${this.score},
      "pos":${this.pos.toString()},
      "dest":${this.dest.toString()},
      "player_weapon":${this.player_weapon},
      "projectiles":[${projectiles_str}],
      "blade":${blade_str}
    }`;
  }
};

class GameState {
  constructor(current_tick = 0, current_round = 0, players = [], coins = []) {
    this.current_tick = current_tick;
    this.current_round = current_round;
    this.players = players;
    this.coins = coins;
  }

  toString() {
    const players_str = this.players.map(p => p.toString()).join(',');
    const coins_str = this.coins.map(c => c.toString()).join(',');

    return `{
      "current_tick":${this.current_tick},
      "current_round":${this.current_round},
      "players":[${players_str}],
      "coins":[${coins_str}]
    }`;
  }
};

export { GameState, PlayerInfo, PlayerWeapon, Coin, Blade, Projectile };
